### ISSUE TITLE
How to use Shared Array Buffer as the wasm memory

### ISSUE THREAD

**Issue by Triuman** [rustwasm/wasm-bindgen#3298]
I would like to run Rust code in a web worker. I am currently passing data between Rust and JS using the Rust's linear memory. Once it's running in the worker, this memory needs to be a shared array buffer.

When initiating the wasm memory manually, we can set the shared property to true.

```JS
const memory = new WebAssembly.Memory({
  initial: 10,
  maximum: 100,
  shared: true,
});
```

How can I do this using wasm-bindgen?

**Comment by Liamolucko** [comment]
You need to pass some flags to `rustc` to enable WebAssembly threading, which will automatically convert the wasm module's memory to a `SharedArrayBuffer`. https://rustwasm.github.io/wasm-bindgen/examples/raytrace.html explains how to do this.

**Comment by Triuman** [comment]
That example made me quite confused. I'm trying to do what they mention in the Caveats section:

> The main thread in a browser cannot block. This means that if you run WebAssembly code on the main thread you can never block, meaning you can't do so much as acquire a mutex. This is an extremely difficult limitation to work with on the web, although one workaround is to run wasm exclusively in web workers and run JS on the main thread. It is possible to run the same wasm across all threads, but you need to be extremely vigilant about synchronization with the main thread.

I'd like to run Rust(wasm) exclusively in a worker. To do this, I thought I'd need to call the wasm init function in a worker and also set the memory to shared. But it seems more complicated than that.

Do you know how to achieve this?

**Comment by Triuman** [comment]
Currently, we have JS running in both main and worker threads. They have a shared array buffer to transfer data from worker to main, and it works without any issues. Since we are moving to Rust for the game logic, we though if we use a `SharedArrayBuffer` for the WASM memory, we can give it to main thread and it can keep reading from it.

So, JS in the main thread will spawn a worker thread which will init WASM. Then, it'll share the WASM memory with the main thread. This would work only if the WASM memory is a `SharedArrayBuffer`.

**Comment by daxpedda** [comment]
It's fine if you use a `SharedArrayBuffer` to transfer data, but you really shouldn't use the WASM memory for that.

It is already used by WASM as it's own memory pool, you can't predict what's gonna happen inside there. You can simply use a separate `SharedArrayBuffer` to transfer data, not re-using the WASM one.

Just to make sure your original question is answered:
> I'd like to run Rust(wasm) exclusively in a worker. To do this, I thought I'd need to call the wasm init function in a worker and also set the memory to shared. But it seems more complicated than that.
> 
> Do you know how to achieve this?

That is exactly what you need to do, but setting the memory to shared is done for you by `wasm-bindgen` as soon as you compile with atomics enabled, see the [raytracing example](https://rustwasm.github.io/wasm-bindgen/examples/raytrace.html).

**Comment by daxpedda** [comment]
> In a demo project, I was able to use WASM memory to transfer Bevy transform data to JS without copying like in the [game of life example.](https://rustwasm.github.io/docs/book/game-of-life/introduction.html) I understand your concerns but we need it to be very efficient. So, we would like to avoid copying as much as possible.

Just to make sure, I was just saying to use a **separate** `SharedArrayBuffer`, this would also avoid copying. You don't have to use the WASM memory to be able to use a `SharedArrayBuffer`.

By the way, as far as I understand, the reason why you only want to run WASM in a worker, is because you are concerned for blocking behavior, but if you just wanted to send data to the worker efficiently, you could just do it with a regular non-blocking channel too.

**Comment by Triuman** [comment]
> Just to make sure, I was just saying to use a **separate** `SharedArrayBuffer`, this would also avoid copying. You don't have to use the WASM memory to be able to use a `SharedArrayBuffer`.

If I use a separate `SharedArrayBuffer`, I still need to copy the values from the Transform components to the `SharedArrayBuffer`, right? -------- After writing this, I realized I actually copy the values right now to be able to reach them from JS.

```rust
    let mut transforms: Vec<Transform> = Vec::with_capacity(query.iter().count());

    for (transform) in query.iter() {
        // I'm new to rust but I must be copying the values here
        transforms.push(*transform);
    }
    let ptr_range = transforms.as_ptr_range();
    // call to JavaScript
    fromRust(ptr_range.start, ptr_range.end);
```

So, what you say is; I can copy these values to a `SharedArrayBuffer` which I could have shared between Rust and JavaScript. Is that correct?

> By the way, as far as I understand, the reason why you only want to run WASM in a worker, is because you are concerned for blocking behavior, but if you just wanted to send data to the worker efficiently, you could just do it with a regular non-blocking channel too.

Yes, I want to put the whole game logic in the worker so that the main thread can deal with rendering without choking. What do you mean by `regular non-blocking channel`?

**Comment by daxpedda** [comment]
> So, what you say is; I can copy these values to a `SharedArrayBuffer` which I could have shared between Rust and JavaScript. Is that correct?

Yes.

> Yes, I want to put the whole game logic in the worker so that the main thread can deal with rendering without choking. What do you mean by `regular non-blocking channel`?

Using [`flume`](https://crates.io/crates/flume) for example, you can wait for a channel without blocking, because it supports both async and sync waiting.

What I'm saying is that you can do the rendering in Rust too on the main thread, [`wgpu`](https://crates.io/crates/wgpu) for example supports WebGL2 and runs fine on the main thread. This would significantly improve (or completely eliminate) any copying or shared memory you have to setup communicating back and forth between JS and Rust.

Basically, the fact that you can't block on the main thread should really not prevent you from running a game (or part of it) in Rust on the main thread. [Bevy](https://bevyengine.org) for example doesn't use workers at all and they do everything in the main thread, which includes rendering, game logic and what not.

**Comment by Triuman** [comment]
That's actually a good point to run Rust in main thread and run threads from Rust and use non-blocking channels. I just learned what a channel means.

We will use Bevy as the game engine actually. We have a custom renderer written in TypeScript and we want to use that alongside with Bevy ECS. The idea is to put Bevy in a worker thread and pass the transforms to the main thread for rendering. We would like to support lower-end mobile devices, so the performance is important and it'll be a multiplayer game. Btw, we already have the game written all in TypeScript. Here is the [link](https://miniverse.mini.com/). (Let me know if it's against the rules here.) Since we will add new features, we want to move it to Rust+Bevy.

It'll take some time to try out the things you said due to other things we need to work on but I'll definitely come here to update. I really appreciate your help!

**Comment by daxpedda** [comment]
Wow, that's so cool, btw this is super off-topic now: but I had some pretty miserable >0.5s input delay.

> It'll take some time to try out the things you said due to other things we need to work on but I'll definitely come here to update. I really appreciate your help!

Happy to help :+1:

**Comment by Triuman** [comment]
The delay is due to a couple of things: the servers are in EU, the client waits until the input reaches to the server to use the input with additional jitter buffer, which makes it really delayed, and using Websockets(TCP). In the next version, we'll make it apply the local input after a a couple of frames like in Rocket League. I still don't know how exactly that should be implemented but I'll try and see. Also, will use Naia library which is communicating through WebRTC data channel.

**Comment by daxpedda** [comment]
Feel free to re-open if you have more questions!
